components/constellation/constellation.rs

```diff
@@ -2084,14 +2120,33 @@ where
         }
     }
 
-    fn handle_load_url_msg(
+    /// Schedule a navigation(via load_url).
+    /// 1: Ask the embedder for permission.
+    /// 2: Store the details of the navigation, pending approval from the embedder.
+    /// 3: In the case of an iframe being navigated, start the navigation via the corresponding event-loop.
+    fn schedule_navigation(
         &mut self,
         top_level_browsing_context_id: TopLevelBrowsingContextId,
         source_id: PipelineId,
         load_data: LoadData,
         replace: bool,
     ) {
-        self.load_url(top_level_browsing_context_id, source_id, load_data, replace);
+        let key = (
+            top_level_browsing_context_id,
+            source_id,
+            load_data.url.clone(),
+        );
+        let pending = self
+            .pending_approval_navigations
+            .entry(key)
+            .or_insert(VecDeque::new());
+        pending.push_back((load_data.clone(), replace));
```

Comment:
Is it possible to have multiple pending navigation for one pipeline? I think it has to be unique. If there's already a pending navigation, fail.